include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -899,55 +899,47 @@ namespace VectorTools
      * MatrixFree implementation of project() for an arbitrary number of
      * components and arbitrary degree of the FiniteElement.
      */
-    template <int components, int fe_degree, int dim, typename VectorType, int spacedim>
-    void project_matrix_free (const Mapping<dim, spacedim>                              &mapping,
-                              const DoFHandler<dim, spacedim>                           &dof,
-                              const ConstraintMatrix                                    &constraints,
-                              const Quadrature<dim>                                     &quadrature,
-                              const Function<spacedim, typename VectorType::value_type> &function,
-                              VectorType                                                &vec_result,
-                              const bool                                                 enforce_zero_boundary,
-                              const Quadrature<dim-1>                                   &q_boundary,
-                              const bool                                                 project_to_boundary_first)
+    template <int components, int fe_degree, int dim, typename Number, int spacedim>
+    void project_matrix_free (const Mapping<dim, spacedim>               &mapping,
+                              const DoFHandler<dim, spacedim>            &dof,
+                              const ConstraintMatrix                     &constraints,
+                              const Quadrature<dim>                      &quadrature,
+                              const Function<spacedim, Number>           &function,
+                              LinearAlgebra::distributed::Vector<Number> &work_result,
```

Comment:
May I suggest to change this back a bit in the direction where it was, namely to use
```
  Function<spacedim, typename LinearAlgebra::distributed::Vector<Number>::value_type> &function
```
in the first line? The point is to make clear that the vector is the choice and the function needs to have the same scalar type. This also helps with compiler messages, because if you have mismatched vector and function arguments, the compiler with the current signature will simply say that no such function exists, whereas if you write as suggested (where the `Number` data type is non-deducible from the function argument) it will say that this is a candidate function but that the function argument can not be converted to the correct data type.